shapely/tests/test_io.py

```diff
@@ -790,6 +864,54 @@ def test_pickle(geom):
     assert_geometries_equal(pickle.loads(pickled), geom, tolerance=0)
 
 
+@pytest.mark.parametrize("geom", all_types_with_z)
+def test_pickle_z(geom):
+    if shapely.get_type_id(geom) == shapely.GeometryType.LINEARRING:
+        pytest.xfail("TODO: Z gets dropped for some reason")
```

Comment:
This is probably because the custom code we have for LinearRings doesn't support more than 2D:

https://github.com/shapely/shapely/blob/60967815a3f4d40c4fcea719380e4287d2bc5e3b/shapely/geometry/polygon.py#L16-L22

We should probably add a helper function in C that converts a LineString to a LinearRing by just getting the CoordSequence from the line and put that in a LinearRing, so that it preserves the full coord seq and its dimensionality.

(doesn't need to happen for this PR)

Reply:
Thanks for the tips, refactoring `get_coordinates` is somewhere down the track.